xtext.cpp

```diff
@@ -0,0 +1,2157 @@
+
+        for ( int i=0; i<=m_length; i++ ) {
+            bool end_of_text = i == m_length;
+
+            // Bidi handling
+            bool bidi_level_changed = false;
+            int last_direction = 1; // LTR if no bidi found
+            if ( m_has_bidi ) {
+                new_bidi_level = i < m_length ? m_bidi_levels[i] : last_bidi_level;
+                if ( i == 0 )
+                    last_bidi_level = new_bidi_level;
+                else if ( new_bidi_level != last_bidi_level )
+                    bidi_level_changed = true;
+                if ( FRIBIDI_LEVEL_IS_RTL(last_bidi_level) )
+                    last_direction = -1; // RTL
+            }
+
+            // Text Unicode script change
+            // Arabic surrounded by hebrew chars would not get its letters joined
+            // if they were all shaped as a single segment. This may probably happen
+            // too with some complex LTR scripts like indic surrounded by latin.
+            // Note: libraqm and Lua library https://github.com/luapower/tr do
+            // a bit more than that by trying to make neutral paired characters part
+            // of a same script segment (_raqm_resolve_scripts(), using a stack, so
+            // probably costly and needing another pass). We don't do that for now.
+            bool script_changed = false;
+            if ( i < m_length ) {
+                hb_script_t script = hb_unicode_script(unicode_funcs, m_text[i]);
+                if ( script != HB_SCRIPT_COMMON && script != HB_SCRIPT_INHERITED && script != HB_SCRIPT_UNKNOWN ) {
+                    if ( prev_script != HB_SCRIPT_COMMON && script != prev_script ) {
+                        m_charinfo[i].flags |= CHAR_SCRIPT_CHANGE;
+                        script_changed = true;
+                        m_has_multiple_scripts = true;
+                    }
+                    prev_script = script;
+                }
+                // Note: as we have here guessed the script of what's to be measured
+                // next, we could store it in m_charinfo (an additional int32...),
+                // so we can pass it to getHbFontData() (or use it and check
+                // ourselves here), to skip fallback fonts that do not support
+                // this script - if fonts announce the scripts they support (I think
+                // I have seen that in some font tables, may be OTF only?).
+            }
+
+            // Line breaking and wrapping
+            bool line_break = false;
+            if ( i == 0 ) {
+                lb_init_break_context(&lbCtx, m_text[i], m_lang ? m_lang : default_lang);
+            }
+            else {
+                // When at end of m_text, add a letter ('Z') so a trailing \n can be
+                // flagged as CHAR_MUST_BREAK_AFTER, so we can show an empty line
+                // and allow the cursor to be positionned after that last \n.
+                int ch = i < m_length ? m_text[i] : 'Z';
```

Comment:
Can't you just shortcut the whole thing and assume brk == CHAR_MUST_BREAK_AFTER?

Reply:
Nope. If last char of file/m_text is "." or "a", I don't want `brk == CHAR_MUST_BREAK_AFTER`. I want it only if the file ends with a `\n` - or one of the rare other unicode chars libunibreak considers the same (like one named "paragraph break").
In the first case, I don't add an empty line in TextEditor at end of file. In the later case, I do.

Reply:
Okay, makes sense ;).